libs/entitlement/qa/src/javaRestTest/java/org/elasticsearch/entitlement/qa/EntitlementsTestRule.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.qa;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.local.PluginInstallSpec;
+import org.elasticsearch.test.cluster.util.resource.Resource;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.yaml.YamlXContent;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+
+class EntitlementsTestRule implements TestRule {
+
+    interface PolicyBuilder {
+        void build(XContentBuilder builder, Path tempDir) throws IOException;
+    }
+
+    final TemporaryFolder testDir;
+    final ElasticsearchCluster cluster;
+    final TestRule ruleChain;
+
+    @SuppressWarnings("this-escape")
+    EntitlementsTestRule(boolean modular, PolicyBuilder policyBuilder) {
+        testDir = new TemporaryFolder();
+        cluster = ElasticsearchCluster.local()
+            .module("entitlement-test-plugin", spec -> setupEntitlements(spec, modular, policyBuilder))
+            .systemProperty("es.entitlements.enabled", "true")
+            .systemProperty("es.entitlements.testdir", () -> testDir.getRoot().getAbsolutePath())
+            .setting("xpack.security.enabled", "false")
+            .build();
+        ruleChain = RuleChain.outerRule(testDir).around(cluster);
+    }
+
+    @Override
+    public Statement apply(Statement statement, Description description) {
+        return ruleChain.apply(statement, description);
+    }
+
+    private void setupEntitlements(PluginInstallSpec spec, boolean modular, PolicyBuilder policyBuilder) {
+        String moduleName = modular ? "org.elasticsearch.entitlement.qa.test" : "ALL-UNNAMED";
+        if (policyBuilder != null) {
+            spec.withEntitlementsOverride(old -> {
+                try {
+                    try (var builder = YamlXContent.contentBuilder()) {
```

Comment:
Why not have this be a String for easier readability?

Reply:
There are variable parts, so a string would need to be a template. A builder seemed easier.